IHP/Job/Dashboard.hs

```diff
@@ -0,0 +1,530 @@
+-- For example, the type 'Include "userId" UpdateUserJob' models an 'UpdateUserJob' type that can access
+-- the 'User' it belongs to through the 'userId' field.
+-- For some reason, GHC doesn't allow us to create implementations of type family applications, so the following doesn't work:
+--
+-- > instance DisplayableJob (Include "userId" UpdateUserJob) where
+--
+-- However, if we wrap this in a concrete type, it works fine. That's what this wrapper is for.
+-- To get the same behavior as above, just define
+--
+-- > instance DisplayableJob (IncludeWrapper "userId" UpdateUserJob) where
+--
+-- and wrap the values as so:
+--
+-- > jobsWithUsers <- query @UpdateUserJob
+-- >    |> fetch
+-- >    >>= mapM (fetchRelated #userId)
+-- >    >>= pure . map (IncludeWrapper @"userId" @UpdateUserJob)
+newtype IncludeWrapper (id :: Symbol) job = IncludeWrapper (Include id job)
+
+-- CREATE TYPE JOB_STATUS AS ENUM ('job_status_not_started', 'job_status_running', 'job_status_failed', 'job_status_succeeded', 'job_status_retry');
+statusToBadge :: JobStatus -> Html
+statusToBadge JobStatusNotStarted= [hsx|<span class="badge badge-info">Not Started</span>|]
+statusToBadge JobStatusRunning = [hsx|<span class="badge badge-info">Running</span>|]
+statusToBadge JobStatusSucceeded = [hsx|<span class="badge badge-success">Succeeded</span>|]
+statusToBadge JobStatusFailed = [hsx|<span class="badge badge-danger">Failed</span>|]
+statusToBadge JobStatusRetry = [hsx|<span class="badge badge-info">Retrying</span>|]
+
+instance (JobsDashboard jobs, AuthenticationMethod authType) => Controller (JobsDashboardController authType jobs) where
+    beforeAction = authenticate @authType
+    action AllJobsAction = autoRefresh $ indexPage @(jobs)
+    action ViewJobAction = autoRefresh $ viewJob @(jobs)
+    action CreateJobAction = newJob @jobs
+    action DeleteJobAction = deleteJob @jobs
+
+-- | We can't always access the type of our job in order to use type application syntax for 'tableName'.
+-- This is just a convinence function for those cases.
+getTableName :: forall job. (DisplayableJob job) => job -> Text
+getTableName _ = tableName @job
+
+data GenericShowView job = GenericShowView job
+instance (DisplayableJob job) => View (GenericShowView job) where
+    html (GenericShowView job) =
+        let
+            table = getTableName job
+        in [hsx|
+            <br>
+                <h5>Viewing Job {get #id job} in {table}</h5>
+            <br>
+            <table class="table">
+                <tbody>
+                    <tr>
+                        <th>Updated At</th>
+                        <td>{get #updatedAt job}</td>
+                    </tr>
+                    <tr>
+                        <th>Created At</th>
+                        <td>{get #createdAt job}</td>
+                    </tr>
+                    <tr>
+                        <th>Status</th>
+                        <td>{statusToBadge (get #status job)}</td>
```

Comment:
in our ihp cloud dashboards we e.g. have a popover that shows the last error if you hover over a failed job:

![image](https://user-images.githubusercontent.com/2072185/112450227-7551da80-8d54-11eb-9297-c970defda030.png)

Could we have this here as well? :) https://getbootstrap.com/docs/4.6/components/popovers/

Reply:
Hey @mpscholten could you send me the HTML + CSS you guys use for these? i'd like to get this in before merge but i'm honestly a bit burnt out right now 😅

Reply:
Yes :)

```haskell
                    renderStatus JobStatusNotStarted = [hsx|<span class="badge badge-secondary">Not Started</span>|]
                    renderStatus JobStatusRunning = [hsx|<span class="badge badge-primary">Running</span>|]
                    renderStatus JobStatusFailed = [hsx|<span class="badge badge-danger" title="Last Error" data-container="body" data-toggle="popover" data-placement="left" data-content={fromMaybe "" (get #lastError job)}>Failed</span>|]
                    renderStatus JobStatusSucceeded = [hsx|<span class="badge badge-success">Succeeded</span>|]
                    renderStatus JobStatusRetry = [hsx|<span class="badge badge-warning" title="Last Error" data-container="body" data-toggle="popover" data-placement="left" data-content={fromMaybe "" (get #lastError job)}>Retry</span>|]
```

```css
.popover-body {
    background-color: #01313f;
    color: rgb(147, 161, 161);
    font-family: Monaco, Menlo, "Ubuntu Mono", Consolas, source-code-pro, monospace;
    font-size: 11px;
}
```

```js
function initPopover() {
    $('[data-toggle="popover"]').popover({ trigger: 'hover click' })
}
$(document).on('ready turbolinks:load', initPopover);
$(initPopover);
```

Reply:
Thank you!!